plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -0,0 +1,699 @@
+			     "Couldn't read initial firehose messages from device");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static guint
+fu_firehose_updater_configure (FuFirehoseUpdater *self, GError **error)
+{
+	gint max_payload_size = CONFIGURE_MAX_PAYLOAD_SIZE_TO_TARGET_IN_BYTES;
+
+	for (int i = 0; i < MAX_CONFIGURE_ATTEMPTS; i++) {
+		gchar *cmd_str;
+		GByteArray *cmd_bytearray = NULL;
+		g_autoptr(XbSilo) rsp_silo = NULL;
+		g_autoptr(XbNode) rsp_node = NULL;
+
+		cmd_str = g_strdup_printf ("<configure"
+					   " MemoryName=\"%s\""
+					   " Verbose=\"%d\""
+					   " AlwaysValidate=\"%d\""
+					   " MaxDigestTableSizeInBytes=\"%d\""
+					   " MaxPayloadSizeToTargetInBytes=\"%d\""
+					   " ZlpAwareHost=\"%d\""
+					   " SkipStorageInit=\"%d\""
+					   "/>",
+					   CONFIGURE_MEMORY_NAME,
+					   CONFIGURE_VERBOSE,
+					   CONFIGURE_ALWAYS_VALIDATE,
+					   CONFIGURE_MAX_DIGEST_TABLE_SIZE_IN_BYTES,
+					   max_payload_size,
+					   CONFIGURE_ZLP_AWARE_HOST,
+					   CONFIGURE_SKIP_STORAGE_INIT);
+		cmd_bytearray = g_byte_array_new_take ((guint8 *)cmd_str, strlen (cmd_str));
+
+		if (!fu_firehose_updater_send_and_receive (self, cmd_bytearray, &rsp_silo, &rsp_node, error)) {
+			g_prefix_error (error, "Failed to run configure command: ");
+			return 0;
+		}
+
+		/* retry if we're told to use a different max payload size */
+		if (!fu_firehose_updater_check_operation_result (rsp_node, NULL)) {
+			guint64 suggested_max_payload_size;
+			g_autoptr(XbNode) root = NULL;
+
+			root = xb_silo_get_root (rsp_silo);
+			suggested_max_payload_size = xb_node_get_attr_as_uint (root, "MaxPayloadSizeToTargetInBytes");
+			if ((suggested_max_payload_size > G_MAXINT) ||
+			    ((gint)suggested_max_payload_size == max_payload_size)) {
+				break;
+			}
+
+			suggested_max_payload_size = max_payload_size;
+			continue;
+		}
+
+		/* if operation is successful, return the max payload size we requested */
+		return max_payload_size;
+	}
+
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "Configure operation failed");
+	return 0;
+}
+
+gboolean
+fu_firehose_updater_reset (FuFirehoseUpdater *self, GError **error)
+{
+	const gchar *cmd_str = "<power value=\"reset\" />";
+	GByteArray *cmd_bytearray = NULL;
+	g_autoptr(XbSilo) rsp_silo = NULL;
+	g_autoptr(XbNode) rsp_node = NULL;
+
+	cmd_bytearray = g_byte_array_append (g_byte_array_new (), (const guint8 *)cmd_str, strlen (cmd_str));
+	if (!fu_firehose_updater_send_and_receive (self, cmd_bytearray, &rsp_silo, &rsp_node, error)) {
+		g_prefix_error (error, "Failed to run reset command: ");
+		return FALSE;
+	}
+
+	if (!fu_firehose_updater_check_operation_result (rsp_node, NULL)) {
+		g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "Reset operation failed");
+		return FALSE;
+	}
+
+	/* read out all of the remaining messages. otherwise modem won't go into reset */
+	while (fu_firehose_updater_send_and_receive (self, NULL, &rsp_silo, &rsp_node, error));
```

Comment:
do we want to have some kind of limit? with by number or by time?

Reply:
added counter